src/Psalm/Internal/Type/SimpleAssertionReconciler.php

```diff
@@ -2501,10 +2516,55 @@ function (Type\Union $t): bool {
             return $existing_var_type;
         }
 
-        $failed_reconciliation = 2;
+        $failed_reconciliation = Reconciler::RECONCILIATION_EMPTY;
 
         return $assertion === 'empty' && $existing_var_type->possibly_undefined
             ? Type::getEmpty()
             : Type::getMixed();
     }
+
+    /**
+     * @param   0|1|2    $failed_reconciliation
+     */
+    private static function reconcileClassConstant(
+        Codebase $codebase,
+        string $class_constant_expression,
+        Union $existing_type,
+        int &$failed_reconciliation
+    ) : Union {
+        if (strpos($class_constant_expression, '::') === false) {
+            return $existing_type;
+        }
+
+        [$class_name, $constant_pattern] = explode('::', $class_constant_expression, 2);
+
+        if (!$codebase->classlike_storage_provider->has($class_name)) {
+            return $existing_type;
+        }
+
+        $constant_regex_pattern = sprintf('#^%s$#', str_replace('*', '.*', $constant_pattern));
```

Comment:
I think a preg_quote could be added somewhere. There's little chance that the provided pattern has a # on it, but I'm not sure what it would do if there is

Reply:
Also, could you extract this feature so it could be applied on other usages of constant resolution elsewhere?

Reply:
Any suggestions on how to name this feature? I really suck when it comes to namings.

Reply:
transformPatternToRegex or something like that?

Reply:
Oh I see, you don't want to extract the whole `type detection` feature abstracted but only the pattern replacement?

Reply:
yeah, it's just to be able to use it where we already resolve the constant but with a simpler method

Reply:
Using `preg_quote` will escape wildcards - I could unescape wildcards afterwards but TBH:
as you say - not sure if thats really necessary and its more complicated to unescape than actually preventing an invalid regex which then simply won't match. I'd say, lets tkae the pill. WDYT @orklah?

Reply:
I've extracted the class-constant type detection into a dedicated, internal, `ClassConstantByWildcardResolver`

Reply:
No worries for preg_quote, we can always go incrementally if we realize later that it's an issue :)